_drafts/best-of-binutils.md

```diff
@@ -0,0 +1,1011 @@
+
+##### Pop Quiz: Which of the Following will Yield Smaller Code?
+
+```c
+const char *test_enum_to_str_v1(eTestEnum e) {
+  switch (e) {
+    case kTestEnum_Val1:
+      return "Val1";
+    case kTestEnum_Val2:
+      return "Val2";
+    case kTestEnum_Val3:
+      return "Val3";
+    case kTestEnum_Val4:
+    default:
+      return "Val4";
+  }
+}
+```
+
+```c
+const char *test_enum_to_str_v2(eTestEnum e) {
+  if (e == kTestEnum_Val1) {
+    return "Val1";
+  } else if (e == kTestEnum_Val2) {
+    return "Val2";
+  } else if (e == kTestEnum_Val3) {
+    return "Val3";
+  } else if (e == kTestEnum_Val4) {
+    return "Val4";
+  }
+  return "Val4";
+}
+```
+
+> NOTE: We've compiled both snipets with `-Os` to force a size based optimization
+
+##### test_enum_to_str_v1 assembly
+
+```
+$ arm-none-eabi-objdump -d nrf52_example/build/nrf52.elf \
+   --disassemble=test_enum_to_str_v1
+0000015c <test_enum_to_str_v1>:
+ 15c:	2802        cmp	r0, #2
+ 15e:	bf9a        itte	ls
+ 160:	4b02        ldrls	r3, [pc, #8]	; (16c <test_enum_to_str_v1+0x10>)
+ 162:	f853 0020   ldrls.w	r0, [r3, r0, lsl #2]
+ 166:	4802        ldrhi	r0, [pc, #8]	; (170 <test_enum_to_str_v1+0x14>)
+ 168:	4770        bx	lr
+ 16a:	bf00        nop
+ 16c:	20000020    .word	0x20000020
+ 170:	2000000a    .word	0x2000000a
+```
+
+##### test_enum_to_str_v2 assembly
+
+```
+$ arm-none-eabi-objdump -d nrf52_example/build/nrf52.elf  \
+   --disassemble=test_enum_to_str_v1
+00000174 <test_enum_to_str_v2>:
+ 174:	b140        cbz	r0, 188 <test_enum_to_str_v2+0x14>
+ 176:	2801        cmp	r0, #1
+ 178:	d008        beq.n	18c <test_enum_to_str_v2+0x18>
+ 17a:	4b05        ldr	r3, [pc, #20]	; (190 <test_enum_to_str_v2+0x1c>)
+ 17c:	4a05        ldr	r2, [pc, #20]	; (194 <test_enum_to_str_v2+0x20>)
+ 17e:	2802        cmp	r0, #2
+ 180:	bf0c        ite	eq
+ 182:	4618        moveq	r0, r3
+ 184:	4610        movne	r0, r2
+ 186:	4770        bx	lr
+ 188:	4803        ldr	r0, [pc, #12]	; (198 <test_enum_to_str_v2+0x24>)
+ 18a:	4770        bx	lr
+ 18c:	4803        ldr	r0, [pc, #12]	; (19c <test_enum_to_str_v2+0x28>)
+ 18e:	4770        bx	lr
+ 190:	20000019    .word	0x20000019
+ 194:	2000000a    .word	0x2000000a
+ 198:	2000000f    .word	0x2000000f
+ 19c:	20000014    .word	0x20000014
+```
+
+Interesting, while both functions have the same behavior, the switch case based approach winds up
+being 50% smaller! It looks like the compiler has optimized that approach better by using a lookup table
+instead of branching.
+
+#### New Feature: visualize jumps
+
+Binutils is constantly adding new features. While writing this post, I discovered a new option in
+`objdump`, `--visualize-jumps`, that was just released in binutils 2.34. This generates
+ascii art showing the branching flow within individual functions which will be very
+helpful for following control flow when stepping through assembly:
+
+```
+$ binutils/objdump  --visualize-jumps=color --disassemble=test_enum_to_str_v2
```

Comment:
Want to use `arm-none-eabi-objdump` here or keep `binutils/objdump`?

Reply:
It was intentional .. it's not yet available in an official arm toolchain